plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs

```diff
@@ -326,14 +326,50 @@ sizeIsAcceptable = \case
   Delay _ t  -> sizeIsAcceptable t
 
 isPure :: Term name uni fun a -> Bool
-isPure = go
+isPure = go True
     where
-        go = \case
-            Var {}      -> True
+        -- See Note [delayAndVarIsPure]
+        go delayAndVarIsPure = \case
+            Var {}                        -> delayAndVarIsPure
             -- These are syntactically values that won't reduce further
-            LamAbs {}   -> True
-            Constant {} -> True
-            Delay {}    -> True
-
+            LamAbs {}                     -> True
+            Constant {}                   -> True
+            Delay _ body                  -> delayAndVarIsPure || go delayAndVarIsPure body
+            -- This case is not needed in PIR's `isPure`, because PIR's beta-reduction pass
+            -- turns terms like this into `Let` bindings.
+            Apply _ (LamAbs _ _ body) arg -> go delayAndVarIsPure arg && go delayAndVarIsPure body
```

Comment:
It's hard for me to fully comprehend what's going on with the whole purity thing, especially with this new addition, but shouldn't it be `go True arg` instead of `go delayAndVarIsPure arg`?

Reply:
Yeah you are right!